evm/src/Hololocker.sol

```diff
@@ -0,0 +1,93 @@
+    event Unlock(address indexed token, address indexed owner, uint256 tokenId, address operator);
+    event Withdraw(address indexed token, address indexed owner, uint256 tokenId, address operator);
+    event LockTimeUpdate(uint256 newValue);
+
+    error InvalidLockTime();
+    error NotUnlockedYet();
+    error TokenNotLocked();
+    error Unauthorized();
+    error UnlockAlreadyRequested();
+
+    // Permits modifications only by the owner of the specified identity.
+    modifier authorized(address token, uint256 tokenId) {
+        _authorized(token, tokenId);
+        _;
+    }
+
+    function lock(address token, uint256 tokenId) external {
+        IERC721(token).transferFrom(msg.sender, address(this), tokenId);
+        nftLockInfo[token][tokenId].owner = msg.sender;
+        nftLockInfo[token][tokenId].operator = msg.sender;
+        emit Lock(token, msg.sender, tokenId, msg.sender);
+    }
+
+    /// @dev Since only authorized user can use this function, it cannot be used without locking NFT beforehand
+    function requestUnlock(address token, uint256 tokenId) external authorized(token, tokenId) {
+        if (nftLockInfo[token][tokenId].unlockTime != 0) {
+            revert UnlockAlreadyRequested();
+        }
+        nftLockInfo[token][tokenId].unlockTime = block.number + lockTime;
+        emit Unlock(token, nftLockInfo[token][tokenId].owner, tokenId, nftLockInfo[token][tokenId].operator);
+    }
+
+    function withdraw(address token, uint256 tokenId) external authorized(token, tokenId) {
+        if (nftLockInfo[token][tokenId].unlockTime == 0 || block.number < nftLockInfo[token][tokenId].unlockTime) {
+            revert NotUnlockedYet();
+        }
+        IERC721(token).transferFrom(address(this), nftLockInfo[token][tokenId].owner, tokenId);
+        emit Withdraw(token, nftLockInfo[token][tokenId].owner, tokenId, nftLockInfo[token][tokenId].operator);
+        delete nftLockInfo[token][tokenId];
+    }
+
+    /// @dev Handles initiating a lock upon direct NFT safeTransferFrom function call
+    function onERC721Received(address operator, address from, uint256 tokenId, bytes calldata)
+        external
+        returns (bytes4)
+    {
+        address token = msg.sender;
+        nftLockInfo[token][tokenId].owner = from;
+        nftLockInfo[token][tokenId].operator = operator;
+        emit Lock(token, from, tokenId, operator);
+        return IERC721Receiver.onERC721Received.selector;
+    }
+
+    function setLockTime(uint256 newLockTime) external onlyOwner {
+        if (newLockTime < MINIMUM_LOCK_TIME) {
+            revert InvalidLockTime();
+        }
+        lockTime = newLockTime;
+        emit LockTimeUpdate(newLockTime);
+    }
+
+    function _authorized(address token, uint256 tokenId) internal view {
+        if (msg.sender != nftLockInfo[token][tokenId].owner && msg.sender != nftLockInfo[token][tokenId].operator) {
```

Comment:
If the "approved address of an NFT" initialized the lock, then we will save it as the "operator", and it will be an authorized user. However, if the NFT was locked by the owner or an operator, we are not saving the "approved address of an NFT" to the authorized users list. This could be changed if necessary, however, as described in the comment below, I wouldn't recommend it.